core/signingrate/threadsafe_signing_rate_tracker.go

```diff
@@ -0,0 +1,363 @@
+package signingrate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/Layr-Labs/eigenda/api/grpc/validator"
+	"github.com/Layr-Labs/eigenda/core"
+)
+
+var _ SigningRateTracker = (*threadsafeSigningRateTracker)(nil)
+
+// Size of the channel buffer for requests to the internal goroutine. This is intentionally sized large in order
+// to absorb bursts of updates. In practice, this won't be necessary unless the number of validators is very large
+// or the batch rate is very high.
+const channelSize = 4096
+
+// A thread-safe wrapper around a SigningRateTracker. Although we could implement this using a mutex,
+// we instead use a goroutine and a channel to serialize access to the underlying SigningRateTracker.
+// This allows operations such as ReportSuccess and ReportFailure to be non-blocking, which is important
+// for performance. These methods are called many times for each batch processed, and we don't want
+// to block the main processing loop on mutex contention.
+type threadsafeSigningRateTracker struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// The base signing rate tracker that does the actual work.
+	base SigningRateTracker
+
+	// Channel for sending requests to the internal goroutine.
+	requests chan any
+}
+
+func NewThreadsafeSigningRateTracker(base SigningRateTracker) SigningRateTracker {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
bike shed: curious your thoughts on an alternate pattern, instead creating a cancelable context inside the constructor, which gets stored as a member variable and called in `Close()`:

You could replace `Close()` with `Start(ctx)`. The input context could be cancelable if needed by the caller, or not if not. Then, you'd use the passed in `ctx` in your control loop, just like you're currently using the existing context.

Change not required, just curious if you think either pattern is preferable over the other

Reply:
There are some pros and cons. Lets discuss this on a call.

Reply:
Fixed as we discussed on the call (use outer context, in tests use t.Context() since it gets auto-cancelled).